stdlib/Base64/src/decode.jl

```diff
@@ -74,17 +91,17 @@ function read_until_end(pipe::Base64DecodePipe, ptr::Ptr{UInt8}, n::UInt)
             unsafe_store!(p + 2, b3 << 6 | b4     )
             p += 3
         else
-            i, p, ended = decode_slow(b1, b2, b3, b4, buffer, i, pipe.io, p, p_end - p, pipe.rest)
+            i, p, ended = decode_slow(UInt8(pipe.decode), b1, b2, b3, b4, buffer, i, pipe.io, p, p_end - p, pipe.rest)
```

Comment:
This is also related to https://github.com/JuliaLang/julia/pull/49849#discussion_r1210090224 (the second method `decodeonechar(x::UInt8, decode::UInt8)`). 
What I observed here was that passing enum value `pipe.decode` (which is of type `Base64Format`) to `decode_slow` causes additional memory allocation and longer run time, while passing it as UInt8 (as done in above code) does not. I observed it with a simple test like the following:
```julia
using BenchmarkTools
import Base64
dectest = "YWJjZGRoYWlvZGJhdWRvaXdwZHBVQVlTQUI="
@benchmark Base64.base64decode(dectest)
```
, while I am not sure about the reason this occurs.

As original Base64 implementation seemed to much concerned on run time performance, I've changed `decode_slow` as in the code above, but I am also wondering to what extent I should be careful on run time performance.

Reply:
Not sure.  Can you revisit this if you change to a 2d array as suggested above?

Reply:
seems 2d array does not affect the result? even though it improves readability.
I am still not sure of the reason for this behavior...
```julia
julia> @benchmark Base64.base64decode(dectest) # current commit with UInt8 conversion
BenchmarkTools.Trial: 10000 samples with 44 evaluations.
 Range (min … max):  870.455 ns … 57.048 μs  ┊ GC (min … max): 0.00% … 94.38%
 Time  (median):     934.091 ns              ┊ GC (median):    0.00%
 Time  (mean ± σ):     1.204 μs ±  1.940 μs  ┊ GC (mean ± σ):  6.89% ±  4.34%

  ▅█▅▂▁▂▂▂▂▁▁          ▁▁▁▁▁▁▁▁                                ▁
  ████████████▇▇▇███████████████▇▇█▇▆▇▆▇▅▆▅▄▃▅▄▅▃▄▄▄▄▅▅▃▃▃▃▄▄▅ █
  870 ns        Histogram: log(frequency) by time      3.12 μs <

 Memory estimate: 2.19 KiB, allocs estimate: 24.

julia> @benchmark Base64.base64decode(dectest) # current commit without the UInt8 conversion
BenchmarkTools.Trial: 10000 samples with 33 evaluations.
 Range (min … max):  881.818 ns … 64.806 μs  ┊ GC (min … max): 0.00% … 94.57%
 Time  (median):     969.697 ns              ┊ GC (median):    0.00%
 Time  (mean ± σ):     1.191 μs ±  1.948 μs  ┊ GC (mean ± σ):  6.05% ±  3.78%

   ▇█▅▃▂▂▂▂▁▁▁                ▁▁▁▁                             ▁
  ▆██████████████▇█▇▇▇▇▇▇▇█████████▇▆▇▆▅▆▅▅▆▅▅▅▄▃▅▄▂▂▃▄▅▃▄▃▂▃▄ █
  882 ns        Histogram: log(frequency) by time      3.07 μs <

 Memory estimate: 2.25 KiB, allocs estimate: 28.

julia> @benchmark Base64.base64decode(dectest) # 2d array with UInt8 conversion at `decode_slow` argument
BenchmarkTools.Trial: 10000 samples with 44 evaluations.
 Range (min … max):  868.182 ns … 48.186 μs  ┊ GC (min … max): 0.00% … 94.56%
 Time  (median):     938.636 ns              ┊ GC (median):    0.00%
 Time  (mean ± σ):     1.228 μs ±  1.755 μs  ┊ GC (mean ± σ):  5.94% ±  4.31%

  ▆█▄▃▂▃▂▂▂▁▁  ▁▁▁▁▁▁▁▁▁▁▁                                     ▁
  ██████████████████████████▇▇▇▇▆▅▆▅▆▅▄▅▅▅▅▅▅▄▅▅▆▄▄▆▄▄▄▅▄▂▅▃▄▄ █
  868 ns        Histogram: log(frequency) by time      3.54 μs <

 Memory estimate: 2.19 KiB, allocs estimate: 24.

julia> @benchmark Base64url.base64decode(dectest) # 2d array without UInt8 conversion
BenchmarkTools.Trial: 10000 samples with 25 evaluations.
 Range (min … max):  900.000 ns … 81.828 μs  ┊ GC (min … max): 0.00% … 95.36%
 Time  (median):     988.000 ns              ┊ GC (median):    0.00%
 Time  (mean ± σ):     1.266 μs ±  2.209 μs  ┊ GC (mean ± σ):  5.62% ±  3.30%

  ▃██▄▃▃▂▂▁▁▁▁     ▁▁▁▁  ▁▁▁▁▁▁▁▁▁▁▁▁▁▁                        ▂
  ██████████████████████████████████████▇▆▇▇▆▅▆▅▄▄▅▅▅▃▄▄▅▄▄▄▁▅ █
  900 ns        Histogram: log(frequency) by time      3.32 μs <

 Memory estimate: 2.25 KiB, allocs estimate: 28.
```

Reply:
I see, I just did not understand well what is regarded as a serious performance regression.
But anyway, it seems that separating what is done in `decode_slow` into two functions (as in 46fed4c) slightly improves performance without the conversion of `decode::Base64Format` into `UInt8`. I think this would be acceptable both from a perspective of performace and readability.